data/src/test/java/org/oppia/android/data/backends/gae/NetworkLoggingInterceptorTest.kt

```diff
@@ -0,0 +1,240 @@
+    mockWebServer.enqueue(MockResponse().setBody(testResponseBody))
+
+    val networkJob = launch {
+      networkLoggingInterceptor.logNetworkCallFlow.collect {
+        assertThat(it.requestUrl).isEqualTo(mockWebServerUrl.toString())
+        assertThat(it.responseStatusCode).isEqualTo(200)
+        assertThat(it.headers).contains(headerString)
+        assertThat(it.body).isEqualTo(testResponseBody)
+      }
+    }
+
+    client.newCall(request).execute()
+    testCoroutineDispatchers.advanceUntilIdle()
+    networkJob.cancel()
+  }
+
+  @Test
+  fun testLoggingInterceptor_makeFailingCallToTopicService_logsNetworkCallFailed() =
+    runBlockingTest {
+      val pageNotFound = HttpURLConnection.HTTP_NOT_FOUND
+
+      val mockResponse = MockResponse()
+        .setResponseCode(pageNotFound)
+        .setBody(testResponseBody)
+
+      mockWebServer.enqueue(mockResponse)
+
+      val networkJob = launch {
+        networkLoggingInterceptor.logNetworkCallFlow.collect {
+          assertThat(it.requestUrl).isEqualTo(mockWebServerUrl.toString())
+          assertThat(it.responseStatusCode).isEqualTo(pageNotFound)
+          assertThat(it.headers).contains(headerString)
+          assertThat(it.body).isEqualTo(testResponseBody)
+        }
+      }
+
+      val failedNetworkJob = launch {
+        networkLoggingInterceptor.logFailedNetworkCallFlow.collect {
+          assertThat(it.requestUrl).isEqualTo(mockWebServerUrl.toString())
+          assertThat(it.responseStatusCode).isEqualTo(pageNotFound)
+          assertThat(it.headers).contains(headerString)
+          assertThat(it.body).isEmpty()
+          assertThat(it.errorMessage).isNotEmpty()
+          assertThat(it.errorMessage).isEqualTo(testResponseBody)
+        }
+      }
+
+      client.newCall(request).execute()
+      testCoroutineDispatchers.advanceUntilIdle()
+      networkJob.cancel()
+      failedNetworkJob.cancel()
+    }
+
+  @Test
+  fun testLoggingInterceptor_makeFailingCallToTopicService_logsNetworkCallFailed_withException() =
+    runBlockingTest {
+      mockWebServer.shutdown()
+
+      val failedNetworkJob = launch {
+        networkLoggingInterceptor.logFailedNetworkCallFlow.collect {
+          assertThat(it.requestUrl).isEqualTo(mockWebServerUrl.toString())
+          assertThat(it.responseStatusCode).isEqualTo(0)
+          assertThat(it.headers).contains(headerString)
+          assertThat(it.body).isEmpty()
+          assertThat(it.errorMessage).isNotEmpty()
+          assertThat(it.errorMessage).contains("Failed to connect to localhost")
+        }
+      }
+
+      try {
+        client.newCall(request).execute()
+      } catch (e: Exception) { }
+
+      testCoroutineDispatchers.advanceUntilIdle()
+      failedNetworkJob.cancel()
+    }
+
+  private fun setUpTestApplicationComponent() {
+    ApplicationProvider.getApplicationContext<TestApplication>().inject(this)
+  }
+
+  private fun setUpRetrofit() {
+    mockWebServer = MockWebServer()
```

Comment:
I don't think this is necessary because the `url` function only creates and returns the passed url with the server route and port number for the mockwebserver included.